### ISSUE TITLE
paddle.fluid.layers.piecewise_decay在输入很长的时候会响应变慢

### ISSUE THREAD

**Issue by bbbxixixixi** [PaddlePaddle/PaddleX#617]
因为paddlex目前没有cyclical cosine annealing decay的实现，所以用piecewise_decay模拟，将每个step和每个step的lr算成list输入piecewise_decay中，总的步数在13000多，所以list长度也13000多。运行后发现速度奇慢。
改成每个epoch手动模拟，单个list长度变成1144左右，这时候能运行了
不知道目前有没有其他的解决方案，每个epoch手动resume checkpoints搞挺麻烦的

**Comment by jiangjiajun** [comment]
如果你是用的Python API训练模型，可以自定义optimizer实现灵活的学习率衰减策略，参考如下示例中，通过自定义optimizer实现piecewise_decay，这种方式下，可根据你的需求调用fluid中的其它衰减策略，如cosine_decay等
```
from paddlex.cls import transforms
import paddlex as pdx

veg_dataset = 'https://bj.bcebos.com/paddlex/datasets/vegetables_cls.tar.gz'
pdx.utils.download_and_decompress(veg_dataset, path='./')

train_transforms = transforms.Compose([
    transforms.RandomCrop(crop_size=224), transforms.RandomHorizontalFlip(),
    transforms.Normalize()
])
eval_transforms = transforms.Compose([
    transforms.ResizeByShort(short_size=256),
    transforms.CenterCrop(crop_size=224), transforms.Normalize()
])

train_dataset = pdx.datasets.ImageNet(
    data_dir='vegetables_cls',
    file_list='vegetables_cls/train_list.txt',
    label_list='vegetables_cls/labels.txt',
    transforms=train_transforms,
    shuffle=True)
eval_dataset = pdx.datasets.ImageNet(
    data_dir='vegetables_cls',
    file_list='vegetables_cls/val_list.txt',
    label_list='vegetables_cls/labels.txt',
    transforms=eval_transforms)

model = pdx.cls.MobileNetV2(num_classes=len(train_dataset.labels))

import paddle.fluid as fluid

# piecewise decay
lr_decay = fluid.layers.piecewise_decay(boundaries=[100, 200], values=[0.1, 0.01, 0.001])
# linear warmup
lr_decay = fluid.layers.linear_lr_warmup(learning_rate=lr_decay, warmup_steps=20, start_lr=0.0, end_lr=0.1)
# optimizer
optimizer = fluid.optimizer.Momentum(lr_decay, momentum=0.9, regularization=fluid.regularizer.L2Decay(1e-04))

model.train(
    num_epochs=10,
    train_dataset=train_dataset,
    train_batch_size=32,
    eval_dataset=eval_dataset,
    optimizer=optimizer,
    save_dir='output/mobilenetv2',
    use_vdl=True)
```

**Comment by bbbxixixixi** [comment]
> 如果你是用的Python API训练模型，可以自定义optimizer实现灵活的学习率衰减策略，参考如下示例中，通过自定义optimizer实现piecewise_decay，这种方式下，可根据你的需求调用fluid中的其它衰减策略，如cosine_decay等
> 
> ```
> from paddlex.cls import transforms
> import paddlex as pdx
> 
> veg_dataset = 'https://bj.bcebos.com/paddlex/datasets/vegetables_cls.tar.gz'
> pdx.utils.download_and_decompress(veg_dataset, path='./')
> 
> train_transforms = transforms.Compose([
>     transforms.RandomCrop(crop_size=224), transforms.RandomHorizontalFlip(),
>     transforms.Normalize()
> ])
> eval_transforms = transforms.Compose([
>     transforms.ResizeByShort(short_size=256),
>     transforms.CenterCrop(crop_size=224), transforms.Normalize()
> ])
> 
> train_dataset = pdx.datasets.ImageNet(
>     data_dir='vegetables_cls',
>     file_list='vegetables_cls/train_list.txt',
>     label_list='vegetables_cls/labels.txt',
>     transforms=train_transforms,
>     shuffle=True)
> eval_dataset = pdx.datasets.ImageNet(
>     data_dir='vegetables_cls',
>     file_list='vegetables_cls/val_list.txt',
>     label_list='vegetables_cls/labels.txt',
>     transforms=eval_transforms)
> 
> model = pdx.cls.MobileNetV2(num_classes=len(train_dataset.labels))
> 
> import paddle.fluid as fluid
> 
> # piecewise decay
> lr_decay = fluid.layers.piecewise_decay(boundaries=[100, 200], values=[0.1, 0.01, 0.001])
> # linear warmup
> lr_decay = fluid.layers.linear_lr_warmup(learning_rate=lr_decay, warmup_steps=20, start_lr=0.0, end_lr=0.1)
> # optimizer
> optimizer = fluid.optimizer.Momentum(lr_decay, momentum=0.9, regularization=fluid.regularizer.L2Decay(1e-04))
> 
> model.train(
>     num_epochs=10,
>     train_dataset=train_dataset,
>     train_batch_size=32,
>     eval_dataset=eval_dataset,
>     optimizer=optimizer,
>     save_dir='output/mobilenetv2',
>     use_vdl=True)
> ```

我说的就是这个，用piecewise_decay模拟
总计需要训练12个epoch，每个epoch1144个step，总的step13728，得出来的boundaries长度13727，values长度13728，这么带入后程序似乎就卡死了，很长时间都没反应
然后我将boundaries和value的长度调整到一个epoch，也就是1144以内，能训练，但是这种手动操作的比较麻烦，要每个epoch重新加载

**Comment by bbbxixixixi** [comment]
这是1000多长度的boundaries的训练耗时，加载完参数，1分钟左右有20step
2021-03-09 13:12:35 [INFO] Load pretrain weights from /home/aistudio/Models/HRNet64/epoch_240.
2021-03-09 13:12:43 [INFO] There are 1531 varaibles in /home/aistudio/Models/HRNet64/epoch_240 are loaded.
2021-03-09 13:13:21 [INFO] [TRAIN] Epoch=1/1, Step=20/1388, loss=0.147148, lr=0.001845, time_each_step=1.84s, eta=0:41:51

这是13000多长度的boundaries
2021-03-09 12:49:23 [INFO] There are 1531 varaibles in /home/aistudio/Models/HRNet64/epoch_211 are loaded.

我等到13.03，依然没有输出后续信息，就中断了，此外构建lr_decay和optimizer也花了几分钟时间，远远超过普通情况
